File: Threading/example1.py
import threading, time

resourceA = threading.Lock()
resourceB = threading.Lock()

def func1():
    
    with resourceA:
        print("Executing func1...")
        time.sleep(0.05)
        print("Fetchin lock b")

        with resourceB:
            print("Got both locks, exec completed")

    print("func1 completed")

def func2():

    with resourceB:
        print("Executing func2...")
        time.sleep(0.05)
        print("Fetchin lock a")

        with resourceA:
            print("Got both locks, exec completed")

    print("func1 completed")

def main():
    
    t1 = threading.Thread(target=func1)
    t2 = threading.Thread(target=func2)

    t1.start()
    t2.start()

    t1.join()
    t2.join()

if __name__ == "__main__":
    main()

File: Threading/example1.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <chrono>

using namespace std;

mutex resourceA;
mutex resourceB;

void func1() {
    unique_lock<mutex> lockA(resourceA);
    cout << "Executing func1..." << endl;
    this_thread::sleep_for(chrono::milliseconds(50));
    cout << "Fetching lock b" << endl;

    unique_lock<mutex> lockB(resourceB);
    cout << "Got both locks, exec completed" << endl;

    cout << "func1 completed" << endl;
}

void func2() {
    unique_lock<mutex> lockB(resourceB);
    cout << "Executing func2..." << endl;
    this_thread::sleep_for(chrono::milliseconds(50));
    cout << "Fetching lock a" << endl;

    unique_lock<mutex> lockA(resourceA);
    cout << "Got both locks, exec completed" << endl;

    cout << "func2 completed" << endl;
}

int main() {
    thread t1(func1);
    thread t2(func2);

    t1.join();
    t2.join();

    cout << "Execution Completed" << endl;
    return 0;
}
